ext/ffi/jit_trampoline.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+
+use crate::NativeType;
+use crate::{tcc::Context, Symbol};
+use std::ffi::c_void;
+use std::ffi::CString;
+use std::fmt::Write as _;
+
+pub(crate) struct Allocation {
+  pub addr: *mut c_void,
+  _ctx: Context,
+  _sym: Box<Symbol>,
+}
+
+macro_rules! cstr {
+  ($st:expr) => {
+    &CString::new($st).unwrap()
+  };
+}
+
+fn native_to_c(ty: &NativeType) -> &'static str {
+  match ty {
+    NativeType::U8 => "uint8_t",
+    NativeType::U16 => "uint16_t",
+    NativeType::U32 => "uint32_t",
+    NativeType::I8 => "int8_t",
+    NativeType::I16 => "uint16_t",
+    NativeType::I32 => "int32_t",
+    NativeType::Void => "void",
+    NativeType::F32 => "float",
+    NativeType::F64 => "double",
+    _ => unimplemented!(),
+  }
+}
+
+pub(crate) fn codegen(sym: &crate::Symbol) -> String {
+  let mut c = String::from("#include <stdint.h>\n");
+  let ret = native_to_c(&sym.result_type);
+
+  // extern <return_type> func(
+  c += "\nextern ";
+  c += ret;
+  c += " func(";
+  // <param_type> p0, <param_type> p1, ...);
+  for (i, ty) in sym.parameter_types.iter().enumerate() {
+    if i > 0 {
+      c += ", ";
+    }
+    c += native_to_c(ty);
+    let _ = write!(c, " p{i}");
+  }
+  c += ");\n\n";
+
+  // void* recv, <param_type> p0, <param_type> p1, ...);
+  c += ret;
+  c += " func_trampoline(";
+  c += "void* recv";
+  for (i, ty) in sym.parameter_types.iter().enumerate() {
+    c += ", ";
+    c += native_to_c(ty);
```

Comment:
It'd be more accurate here to use a different type mapping for the trampoline, since V8 will never actually call with u8s, u16s etc. All of those parameter types should become 32 bit numbers, and corresponding casting & masking should be applied to convert the parameter value to the actual FFI call parameter value.

Reply:
V8's Fast API supported types can be referenced here: https://v8docs.nodesource.com/node-18.2/da/d05/classv8_1_1_c_type_info.html (This is just a very rough API doc, it doesn't really explain much. Not all of the types are also supported in all positions.)

As evident, only 32bit and 64bit integers are mentioned.

Beyond that, u8 / u32 casting and masking probably needs its own reference :)